### PR TITLE
Post categories

### PR DESCRIPTION
**Pull request by tux255**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,16 +1,48 @@
 version: 2.1
 orbs:
-  ruby: circleci/ruby@0.1.2
+  ruby: circleci/ruby@1.0
+  node: circleci/node@2
+  rails: finalcad/rails@2.1.0
```
Comment (igor04) [inline .circleci/config.yml:5]: you have CI changes but for some reason your tests still failing

File: Gemfile [hunk 2]
```diff
@@ -36,6 +36,7 @@ group :development, :test do
   gem 'guard', '~> 2.17'
   gem 'guard-rspec', require: false
   gem 'pry-rails', '~> 0.3.9'
+  gem 'rails-controller-testing'
```
Comment (igor04) [inline Gemfile:39]: I'm not sure why this is needed

File: app/controllers/admin/categories_controller.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      authorize category
+    end
+
+    def new
+      @category = Category.new
+      authorize @category
+    end
+
+    def create
+      @category = current_user.categories.new(category_params)
```
Comment (igor04) [inline app/controllers/admin/categories_controller.rb:22]: is there a reason to have relation between user and categories?

File: app/controllers/admin/categories_controller.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        render 'new'
+      end
+    end
+
+    def edit
+      @category = Category.find(params[:id])
+    end
+
+    def update
+      @category = current_user.categories.new(params[:post])
```
Comment (igor04) [inline app/controllers/admin/categories_controller.rb:37]: not sure but it looks like you are creating a new category instead of updating existed

File: app/controllers/admin/categories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
+module Admin
+  class CategoriesController < Admin::BaseController
+    def index
+      @categories = policy_scope(Category)
+      authorize @categories
```
Comment (igor04) [inline app/controllers/admin/categories_controller.rb:5]: not sure that we need it for admin namespace

File: app/controllers/admin/posts_controller.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    # Use callbacks to share common setup or constraints between actions.
+    def set_post
+      @post = Post.find(params[:id])
+      authorize @post
+    end
+
+    # Only allow a list of trusted parameters through.
+    def post_params
+      params.require(:post).permit(:title, :body, :user_id)
```
Comment (igor04) [inline app/controllers/admin/posts_controller.rb:79]: not sure why do you need `user_id` as a parameter, because it looks the same as `current_user.id`

File: app/models/category.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+class Category < ApplicationRecord
+  belongs_to :user
+  has_many :subcategories, class_name: 'Category', foreign_key: 'parent_id', dependent: :destroy
+  belongs_to :parent_category, class_name: 'Category', optional: true
+  has_many :posts, foreign_key: 'category_id'
```
Comment (igor04) [inline app/models/category.rb:5]: foreign_key is unnecessary

File: app/models/category.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class Category < ApplicationRecord
+  belongs_to :user
+  has_many :subcategories, class_name: 'Category', foreign_key: 'parent_id', dependent: :destroy
+  belongs_to :parent_category, class_name: 'Category', optional: true
```
Comment (igor04) [inline app/models/category.rb:4]: I think that you need to specify `primary_key:  :parent_id`  to make this relation works properly

File: app/policies/category_policy.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
+class CategoryPolicy < ApplicationPolicy
+  def index?
```
Comment (igor04) [inline app/policies/category_policy.rb:2]: it looks like an unnecessary policy because only the admin has access to it and admin namespace has restrictions

File: app/views/admin/categories/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+<%= form_for [:admin,@category] do |f| %>
```
Comment (igor04) [inline app/views/admin/categories/edit.html.erb:1]: space is needed after `,`

File: app/views/admin/categories/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,4 @@
+<% @categories.each do |cat| %>
+  <%= cat.title %>
+  <%= cat.id %>
+<% end unless @categories.blank? %>
```
Comment (igor04) [inline app/views/admin/categories/index.html.erb:4]: it's better to avoid using `unless` with multiline expressions, I think in the current case there is no need for such condition at all

File: app/views/admin/categories/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<% content_for(:add_item_button) { link_to 'Add Category', new_admin_category_path, class: 'btn btn-primary' } %>
```
Comment (igor04) [inline app/views/admin/categories/show.html.erb:1]: please use do end block instead `<% content_for(:add_item_button) do %> ....`

File: app/views/admin/users/_user.html.erb [hunk 13]
```diff
@@ -10,7 +10,7 @@
   <%= form_for([:admin,user]) do |f| %>
     <%= f.select(:role, User.roles.keys.map {|role| [role.titleize,role]}) %>
     <%= f.submit 'Change Role', :class => 'button-xs' %>
-  <% end %>
+  <% end unless user == current_user %>
```
Comment (igor04) [inline app/views/admin/users/_user.html.erb:13]: use `if` instead

File: spec/factories/post.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
+require 'faker'
+
+FactoryBot.define do
+  factory :post do
+    title { Faker::Lorem.sentence(word_count: 4) }
+    body { Faker::Lorem.sentences(number: 5) }
+
+    trait :with_no_body do
+      body { '' }
```
Comment (igor04) [inline spec/factories/post.rb:9]: instead of `FactoryBot.create(:post, :with_no_body)` you could simply have `FactoryBot.create(:post, body: '')`, so not sure why this traits are needed

File: spec/models/category_spec.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    it 'save' do
+      expect(valid_category.save).to be_truthy
+    end
+  end
+
+  context 'when create subcategory' do
+    it 'is created' do
+      sub_title = 'Subcategory title'
+      cat = user.categories.create!(title: 'Main category')
+      subcat = cat.subcategories.new(title: sub_title, user_id: user.id)
```
Comment (igor04) [inline spec/models/category_spec.rb:38]: not sure why categories need relation to user

File: spec/models/survey_spec.rb [hunk 16]
```diff
@@ -1,36 +1,95 @@
...
+  describe 'survey validation test' do
+    let(:user) { FactoryBot.create(:user, :admin) }
 
-  it 'is not valid with less than two answers' do
-    survey = Survey.new(question: 'Is this thing works?', start_date: Date.today, end_date: Date.tomorrow)
-    survey.survey_answers.new(answer: 'single answer')
-    expect(survey).to_not be_valid
-  end
+    it 'is valid with valid attributes' do
+      expect(user.surveys.new(
```
Comment (igor04) [inline spec/models/survey_spec.rb:8]: it could be `let(:survey) { user.surveys.new(question: question....` and `let(:question) { 'Question?' }` so when you need test case with blank question you will just define `let(:question) { '' }` and will not touch or duplicate `survay`

File: spec/requests/posts_spec.rb [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    it '#destroy' do
+      new_post = user.posts.new(
+        title: 'post title',
+        body: 'body content'
+      )
+      expect(new_post.save).to be_truthy
+      expect(new_post.destroy).to be_truthy
+    end
+
+    it 'persisted' do
```
Comment (igor04) [inline spec/requests/posts_spec.rb:52]: this test case looks like has nothing related to `requests`

File: spec/requests/posts_spec.rb [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+    end
+
+    it '#update' do
+    end
+
+    it '#destroy' do
+      new_post = user.posts.new(
+        title: 'post title',
+        body: 'body content'
+      )
```
Comment (igor04) [inline spec/requests/posts_spec.rb:47]: record need to be created with factory bot and then `delete :posts, params: { id: record.id }` used

File: spec/requests/survey_spec.rb [hunk 19]
```diff
@@ -1,7 +1,59 @@
...
+
+    it '#create' do
+      expect do
+        post '/admin/surveys',
+             params: { survey: survey_atts }
+      end.to change { Survey.all.size }.by(1)
+    end
+
+    it '#edit' do
+      survey = admin.surveys.create(survey_atts)
```
Comment (igor04) [inline spec/requests/survey_spec.rb:53]: it expectst to be `FactoryBot.create(...)`

File: .circleci/config.yml (hunk 1 above)
Comment (tux255) [reply .circleci/config.yml:5]: this is true. For now I do not know how to fix it. Will need your help with this one.

File: Gemfile (hunk 2 above)
Comment (tux255) [reply Gemfile:39]: removed

File: app/controllers/admin/categories_controller.rb (hunk 3 above)
Comment (tux255) [reply app/controllers/admin/categories_controller.rb:22]: there is no exact reason.
I thought it is good idea to know which user created category.
Do you think different?

File: app/controllers/admin/categories_controller.rb (hunk 4 above)
Comment (tux255) [reply app/controllers/admin/categories_controller.rb:37]: found and resolved in other branch

File: app/controllers/admin/categories_controller.rb (hunk 5 above)
Comment (tux255) [reply app/controllers/admin/categories_controller.rb:5]: got it. solved by removing policy_scope

File: spec/requests/posts_spec.rb (hunk 17 above)
Comment (tux255) [reply spec/requests/posts_spec.rb:52]: got it will replace with delete request

File: app/models/category.rb (hunk 8 above)
Comment (tux255) [reply app/models/category.rb:4]: Do you mean like this: `has_many :subcategories, class_name: 'Category', primary_key: :parent_id, dependent: :destroy`

if I do like this it is not working. Error: `SQLite3::SQLException: no such column: categories.category_id`

File: app/views/admin/categories/index.html.erb (hunk 11 above)
Comment (tux255) [reply app/views/admin/categories/index.html.erb:4]: it gives error if @categories are empty

File: app/models/category.rb (hunk 8 above)
Comment (tux255) [reply app/models/category.rb:4]: understood what you mean. updated code. it was working and keeps working
